src/python/pants/backend/jvm/tasks/coursier_resolve.py

```diff
@@ -0,0 +1,478 @@
+          pickle.dump(compile_classpath.get_artifact_classpath_entries_for_targets([t]), f)
+        vt.update()
+
+  def _load_result_from_cache(self, compile_classpath, all_vts, target_resolution_filename):
+    """
+
+    :param compile_classpath:
+    :param all_vts:
+    :param target_resolution_filename:
+    :return: True if success; False if any of the classpath is not valid anymore.
+    """
+    temp_store = []
+
+    for vt in all_vts:
+      t = vt.target
+      if isinstance(t, JarLibrary):
+        # compile_classpath.add_jars_for_targets
+        with open(os.path.join(vt.results_dir, target_resolution_filename), 'r') as f:
+          tuples_conf_artifact_classpath = pickle.load(f)
+          for conf, artifact_classpath in tuples_conf_artifact_classpath:
+            if not os.path.exists(artifact_classpath.path) \
+                or not os.path.exists(artifact_classpath.cache_path) \
+                or not os.path.exists(os.path.realpath(artifact_classpath.cache_path)):
+              print("cache verification failed")
+              return False
+            temp_store.append((t, tuples_conf_artifact_classpath))
+
+    # If all artifacts path are valid, add them to compile_classpath
+    for t, tuples_conf_artifact_classpath in temp_store:
+      compile_classpath.add_elements_for_target(t, tuples_conf_artifact_classpath)
+    print("cache verified")
+    return True
+
+  def _construct_cmd_args(self, classified_jars, classifier, common_args, global_excludes, local_exclude_args,
+                          pinned_coords, coursier_workdir):
+    cmd_args = list(common_args)
+    if classifier:
+      cmd_args.extend(['--classifier', classifier])
+    for j in classified_jars:
+      if not j.rev:
+        continue
+
+      if j.intransitive:
+        cmd_args.append('--intransitive')
+
+      cmd_args.append(j.coordinate.simple_coord)
+
+      # Force requires specifying the coord again with -V
+      if j.force:
+        cmd_args.append('-V')
+        cmd_args.append(j.coordinate.simple_coord)
+
+    # Force pinned coordinates
+    for j in pinned_coords:
+      cmd_args.append('-V')
+      cmd_args.append(j.simple_coord)
+
+    if local_exclude_args:
+
+      with temporary_file(coursier_workdir, cleanup=False) as f:
+        exclude_file = f.name
+        with open(exclude_file, 'w') as ex_f:
+          ex_f.write('\n'.join(local_exclude_args).encode('utf8'))
+
+        cmd_args.append('--soft-exclude-file')
+        cmd_args.append(exclude_file)
+
+    # TODO(wisechengyi): support exclusion on the whole org
+    for ex in global_excludes:
+      if ex.org and ex.name:
+        cmd_args.append('-E')
+        cmd_args.append('{}:{}'.format(ex.org, ex.name))
+    return cmd_args
+
+  @classmethod
+  def _flatten_resolution_by_root(cls, result):
+    """
+    Flatten the resolution by root dependencies. If we want to resolve X and Y, and X->A->B, and Y -> C,
```

Comment:
Will review once the `self.invalidated` block is clearer, but: it looks like this means you're doing one resolution per root? Neat. That will require an interesting cache strategy I think. But I'm pretty sure you'll need to set `invalidate_dependents=True` in order to have a root re-resolved if its dependencies changed.

Also, I don't... really understand how this would work, because we don't support re-compiling targets multiple times in one run, and if one root wants finagle at version X, while some other root wants it at version Y, you'd have a dilemma.

Reply:
no flattening is needed since it is done by coursier now. https://github.com/coursier/coursier/pull/692 shows the format.

invalidate scheme is now
1. do `invalidated` on each target
2. use the produced `invalidation_check` to create a versioned target set.